apkenv.c

```diff
@@ -227,6 +243,82 @@ install_overrides(struct SupportModule *module)
     }
 }
 
+void
+dalvikvm_JNIEnv_DeleteLocalRef(JNIEnv* p0, jobject p1)
+{
+#ifdef APKENV_DEBUG
+    printf("dalvikvm_JNIEnv_DeleteLocalRef(%x)\n", p1);
+#endif
+    global.dalvik_copy_env.DeleteLocalRef(p0, p1);
+    (*p0)->ExceptionClear(p0);
+}
+
+
+static void
+install_default_dalvik_overrides()
+{
+    struct JNINativeInterface *dalvik_env = (struct JNINativeInterface *)global._env->functions;
+
+    int pagesize = sysconf(_SC_PAGESIZE);
+    mprotect((char*)((int)dalvik_env & ~(pagesize-1)), pagesize*2, PROT_READ | PROT_WRITE);
+
+    dalvik_env->NewGlobalRef = JNIEnv_NewGlobalRef;
+    dalvik_env->DeleteGlobalRef = JNIEnv_DeleteGlobalRef;
+
+    dalvik_env->GetObjectClass = JNIEnv_GetObjectClass;
+
+    dalvik_env->FindClass = JNIEnv_FindClass;
+    dalvik_env->GetMethodID = JNIEnv_GetMethodID;
+    dalvik_env->GetFieldID = JNIEnv_GetFieldID;
+    dalvik_env->GetStaticFieldID = JNIEnv_GetStaticFieldID;
+    dalvik_env->GetStaticObjectField = JNIEnv_GetStaticObjectField;
+    dalvik_env->GetStaticMethodID = JNIEnv_GetStaticMethodID;
+    dalvik_env->GetObjectField = JNIEnv_GetObjectField;
+
+    dalvik_env->DeleteLocalRef = dalvikvm_JNIEnv_DeleteLocalRef;
+
+    dalvik_env->RegisterNatives = JNIEnv_RegisterNatives;
+
+    mprotect((char*)((int)dalvik_env & ~(pagesize-1)), pagesize*2, PROT_READ);
+}
+
+static void
+install_dalvik_overrides(struct SupportModule *module)
+{
+    int i;
+    void **dalvik, **override;
+
+    int pagesize = sysconf(_SC_PAGESIZE);
+
+    /* Install overrides for JNIEnv */
+    dalvik = (void**)(global._env->functions);
+    mprotect((char*)((int)dalvik & ~(pagesize-1)), pagesize*2, PROT_READ | PROT_WRITE);
```

Comment:
Same comments as above. If we do that in multiple places around this file, maybe it makes sense to write a `mprotect()` wrapper that takes care of page-aligning the pointer and calculating the right size (pseudocode):

```
void set_writeable(char *addr, size_t len, bool writeable)
{
    char *offset = addr % sysconf(_SC_PAGESIZE);
    mprotect(addr - offset, offset + len, writable ? (PROT_READ | PROT_WRITE) : (PROT_READ));
}
```